pkg/metrics/collector/netlink_metrics.go

```diff
@@ -14,84 +16,318 @@ limitations under the License.
+	defer curMap.mapMux.Unlock()
+	curMap.ipMap[key] = value
+}
+
+func (curMap *safeIPMap) safeIPDelete(key string) {
+	curMap.mapMux.Lock()
+	defer curMap.mapMux.Unlock()
+	delete(curMap.ipMap, key)
+}
+
+func (curMap *safeIPMap) safeIPRead(key string) (*v1.Pod, bool) {
+	curMap.mapMux.RLock()
+	defer curMap.mapMux.RUnlock()
+	podDef, ok := curMap.ipMap[key]
+	return podDef, ok
+}
+
+// Functionality for pod watching.
+func createPodWatch() error {
+	// Creates the in-cluster config.
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	watchlist := cache.NewListWatchFromClient(
+		clientset.CoreV1().RESTClient(),
+		"pods",
+		v1.NamespaceAll,
+		fields.OneTermEqualSelector("spec.nodeName", nodeName),
+	)
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    onPodAdd,
+			UpdateFunc: onPodUpdate,
+			DeleteFunc: onPodDelete,
+		},
+	)
+
+	stopper := make(chan struct{})
+
+	go controller.Run(stopper)
+
+	if !cache.WaitForCacheSync(stopper, controller.HasSynced) {
+		glog.Infof("Timed out waiting for caches to sync.")
+	}
+
+	return nil
+}
+
+func hasPodIP(pod *v1.Pod) bool {
+	// Pod not assigned.
+	if pod.Status.PodIP == "" {
+		return false
+	}
+	// Pod running on host network.
+	if pod.Status.PodIP == pod.Status.HostIP {
+		return false
+	}
+	return true
+}
+
+func onPodAdd(podObj interface{}) {
+	pod := podObj.(*v1.Pod)
+	if hasPodIP(pod) {
+		ipMap.safeIPWrite(pod.Status.PodIP, pod)
+	}
+}
+
+func onPodUpdate(oldPod, newPod interface{}) {
+	prevPod := oldPod.(*v1.Pod)
+	pod := newPod.(*v1.Pod)
+	if prevPod.Status.PodIP == pod.Status.PodIP {
+		return
+	}
+	if hasPodIP(prevPod) {
+		ipMap.safeIPDelete(prevPod.Status.PodIP)
+	}
+	if hasPodIP(pod) {
+		ipMap.safeIPWrite(pod.Status.PodIP, pod)
+	}
+	glog.Infof("pods are: %+q", ipMap.ipMap)
+}
+
+func onPodDelete(podObj interface{}) {
+	pod := podObj.(*v1.Pod)
+	if hasPodIP(pod) {
+		ipMap.safeIPDelete(pod.Status.PodIP)
+	}
+}
+
+func updatePodStats(stats netlinkStats, tcpInfo *tcp.LinuxTCPInfo) netlinkStats {
+	stats.tcpConnections++
+	if tcpInfo == nil {
+		return stats
+	}
+	stats.retransmits += uint64(tcpInfo.Retrans)
+	return stats
+}
+
+func createStatMap(snapshots []*parser.Snapshot) map[*v1.Pod]netlinkStats {
+	statMap := make(map[*v1.Pod]netlinkStats)
+	for _, snapshot := range snapshots {
+		if snapshot.SockInfo == nil {
+			continue
+		}
+		if pod, ok := ipMap.safeIPRead(snapshot.SockInfo.SrcIP); ok {
+			statMap[pod] = updatePodStats(statMap[pod], snapshot.TCPInfo)
+		}
+
+		if pod, ok := ipMap.safeIPRead(snapshot.SockInfo.DstIP); ok {
+			statMap[pod] = updatePodStats(statMap[pod], snapshot.TCPInfo)
+		}
+	}
+	return statMap
+}
+
+func NewNetlinkCollector() (Collector, error) {
+	return &netlinkCollector{}, nil
+}
+
+func getNamespaces() ([]netns.NsHandle, error) {
```

Comment:
Just double checking - if we disable pod watch and don't enable HostPID fo netd, would it still be able to expose the node level metrics properly (I guess yes since we will be extracting the host network namespace from netd its own's PID)?

Reply:
Hmm, there may be some issues because we may not count all connections on every namespace. There may be some undercounting involved.

Reply:
Got it - do you mind filing an issue so that we don't lost track of this? Then we could merge this PR.

Reply:
Hmm. I think we can do two things - always collect metrics for the root ns (and label it by the node IP I guess) and optionally if the pod watch is enabled, also collect metrics for the pod namespaces.

I think this will require some additional work to figure out how to export the root ns metrics into a K8sContainer schema. So for now, I think just create an issue so we don't lose track and we can merge this PR.